ops/log.py

```diff
@@ -53,6 +55,20 @@ def setup_root_logging(
         debug: if True, write logs to stderr as well as to juju-log.
         exc_stderr: if True, write uncaught exceptions to stderr as well as to juju-log.
     """
+    logging.captureWarnings(True)
+
+    def custom_warning_formatter(
+        message: typing.Union[str, Warning],
+        category: typing.Type[Warning],
+        filename: str,
+        lineno: int,
+        line: typing.Optional[str] = None,
```

Comment:
This is tangential to the PR, but I would have thought ruff would complain here about `line` not being used, and require a pragma to ignore it. But I guess it is not doing that.